### PR TITLE
WIP /elections endpoint

### PR DESCRIPTION
**Pull request by chris48s**
**Stuff I have done:**
* Expose `/elections/` and `/elections/<election_id>` endpoints via devs.DC API
* Remove or reformat any links returned by EE so that we don't link out to http://elections.democracyclub.org.uk/
* Expose sensible set of filter/query params on `/elections/`
* Link to `/elections/<election_id>` from ballot objects returned by `/postcode`, `/address` instead of allowing query by postcode centroid on `/elections/`
* Bit of tidy up/refactoring of existing code

**TODO to finish this PR:**
* Tests: In general, this whole API Client layer lacks tests (refs #6) so I want to get tests written for all of that next week as well as just the new code I've added here. This later of the application deals with all the communication with the outside world, so there are a lot of external interactions to mock out.

**TODO in near future (but not needed to merge this bit):**
* Sandbox outputs (this is less important because `/elections/` always has real data)
* Documentation for `/elections/` endpoint
* There are some changes which I would like to move upstream to the EE API but impact on YNR/WCIVF, so I'm going to put them on hold temporarily and fiddle them at the proxy layer for now

Finally, one thing I realise as I think about documenting the API outputs is that the way we identify a ballot is not very nice. At the moment `"group_type": null` is the thing that tells you if an election is a ballot or a group. As an outsider, it feels like a bit of a strange way to expose quite a key bit of info though. What do you think of changing that to `"group_type": "ballot"` or adding an extra key like `"ballot": true/false`?

### CODE REVIEW

File: aggregator/apps/api/v1/api_client.py [hunk 1]
```diff
@@ -92,11 +107,86 @@ def get_data_for_address(self, slug):
...
+        election.pop("tmp_election_id", None)
+
+        if "organisation" in election and election["organisation"]:
+            election["organisation"].pop("url", None)
+
+        if "division" in election and election["division"]:
+            election["division"].pop("geography_curie", None)
+            election["division"].pop("mapit_generation_low", None)
+            election["division"].pop("mapit_generation_high", None)
+            election["division"]["divisionset"].pop("mapit_generation_id", None)
```
Comment (chris48s) [inline aggregator/apps/api/v1/api_client.py]: There are some of these things I actually want to completely remove in EE, but doing so may impact on YNR/WCIVF so for the moment I'm just going to strip them here. I'm using `.pop("field_name", None)` here because it means that we will remove `"field_name"` from the dict if it exists but it won't throw an exception if it doesn't. This makes it a bit easier to co-ordinate because it means I can make those changes in EE without breaking our code here and then tidy up here later.

File: aggregator/apps/api/v1/api_client.py [hunk 2]
```diff
@@ -92,11 +107,86 @@ def get_data_for_address(self, slug):
...
+
+        There are several things this acheives:
+        1. There are some params that the devs.DC API accepts
+           that we don't want to pass through to the EE API (e.g: auth_token)
+        2. There are some params supported by the EE API
+           that we don't want to expose to dev.DC users (e.g: postcode)
+        3. In future, if we wanted to pass additional params to EE
+           (e.g: a tracking param or API key, like we do with WDIV),
+           they'll come back reflected in the next/previous URLs.
+           Cleaning prevents us exposing those params to the devs.DC user.
```
Comment (chris48s) [inline aggregator/apps/api/v1/api_client.py]: I think I've explained this OK here, but just to clarify what this third point with an example:

With WDIV and WCIVF, we pass `utm_medium=devs.DC API` in the querystring for stats purposes because WDIV and WCIVF store that in the `LoggedPostcode` table. I haven't bothered with EE because EE doesn't log it, but if we did that, the `next` URL in the pagination links will come back as `https://elections.democracyclub.org.uk/api/elections/?limit=100&offset=100&utm_medium=devs.DC API`. Same would go for an API key, or whatever. This allows us to safely modify the querystring at this layer before we forward it on to EE without it impacting on the user. We'd just expose the pagination link as `https://developers.democracyclub.org.uk/api/v1/elections/?limit=100&offset=100`

File: aggregator/apps/api/v1/api_client.py (hunk 1 above)
Comment (symroe) [reply aggregator/apps/api/v1/api_client.py]: Agree with the approach – FWIW, YNR/WCIVF don't use this anywhere, so it's safe to remove from EE now I reckon.

File: aggregator/apps/api/v1/api_client.py (hunk 2 above)
Comment (symroe) [reply aggregator/apps/api/v1/api_client.py]: 👍

File: aggregator/apps/api/v1/api_client.py (hunk 1 above)
Comment (chris48s) [reply aggregator/apps/api/v1/api_client.py]: There is one change I need to make, which is switching YNR to use `official_identifier` instead of `geography_curie`:

https://github.com/DemocracyClub/yournextrepresentative/blob/69caa54c3d5326860852298a892048b51b0aa7ac/ynr/apps/elections/uk/every_election.py#L137-L141

`geography_curie` is just an alias for `official_identifier` now though, so there's not really any point exposing both of them in the API - its just confusing to have the same value with 2 different field names.

https://github.com/DemocracyClub/EveryElection/blob/e936ef13168e3a7fb7ac283e4c006e45ba8e43f2/every_election/apps/organisations/models/divisions.py#L115-L117

File: aggregator/apps/api/v1/api_client.py [hunk 3]
```diff
@@ -92,11 +107,86 @@ def get_data_for_address(self, slug):
...
+        """
+        allowed_query_params = [
```
[the remark this answers is not in the document]
Comment (chris48s) [reply aggregator/apps/api/v1/api_client.py]: Yep. Whether you block stuff you don't want or allow stuff you do want, we have to remember to update your list under some situation.

Essentially the tradeoff is:

Allow list:
If we add a new param to the EE API which we want to expose to users, we've got to remember to add it to the allow list when we update the documentation over here.

Block list:
If we start passing anything extra between devs.DC and EE internally (e.g: an API key), we have to remember to add it to the deny list, otherwise the default position is its going to be automatically reflected to the user.

I did think about both ways, but I think the first option is safer in that respect. Also, explicitly stating what params can be passed also serves as a reference for what should be documented.

File: aggregator/apps/api/v1/api_client.py [hunk 4]
```diff
@@ -22,31 +25,50 @@ def get_event_loop():
...
     timeout = aiohttp.ClientTimeout(total=10)
     async with aiohttp.ClientSession(timeout=timeout) as session:
         async with session.get(url, params=params, headers=headers) as response:
             body = await response.json()
             return {"url": url, "json": body, "status": response.status}
 
 
-class ApiClient:
-    def __init__(self):
+def proxy_single_request(loop, url, params=None):
```
[the remark this answers is not in the document]
Comment (chris48s) [reply aggregator/apps/api/v1/api_client.py]: If we want to do that (and maybe throw something like a `400 Bad Request` instead of just the standard `404 Not Found`?) if you try and pass an invalid ID, I'd prefer to do that in the EE API and forward the error response.

I reckon as far as possible, we should try and keep "domain knowledge" (e.g: what is/isn't a valid election id) in the upstream APIs and keep this layer fairly "dumb" in that respect.

File: aggregator/apps/api/v1/api_client.py (hunk 1 above)
Comment (symroe) [reply aggregator/apps/api/v1/api_client.py]: Ah, sorry, I was talking about the mapit stuff 👍

File: aggregator/apps/api/v1/api_client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (symroe) [reply aggregator/apps/api/v1/api_client.py]: 👍

**Comment by chris48s** [comment]
> I'd be interested to know what you plan to do with mocking the remote responses. FWIF I've found vcr to be really useful for quick and dirty testing, but too hard to update the fixtures. Adding a single new field to a response is almost impossible when the data is JSON squashed in to a YAML string. I reckon just keeping JSON files with the various responses is much easier in the long run.

I fiddled a bit with both approaches. tbh, with this project I'd already gone down the route of just storing local copies of JSON responses, so I've doubled down on it here. With the `/postcode` and `/address` tests I already doubled up the sandbox outputs as expected values for the tests to DRY up and help us keep stuff in sync: if the sandbox outputs don't match the results of running the code, the tests will fail too. We need a bit of code to manage loading the inputs/expected responses and the process of injecting the mocks deep in the call-stack is a bit ugly, but VCR/betamax/similar solve a particular problem that isn't quite the shape we've got here.

In general I think those tools are really useful in the case where you can hit an upstream API and get back the exact response you want to test against reliably, but you don't want your test to rely on an external interaction for stability/performance reasons. https://elections.democracyclub.org.uk/api/elections/local.surrey-heath.st-michaels.2019-05-02.json would be an example of this. If I call that today and then again next year, I'd expect the exact same thing to come back both times (or at least, until we change something :wink: )

These tools are less good if the reason for mocking is that your upstream API's content changes over time and you need to mock so you can nail down a particular output which is returned consistently. Something like https://wheredoivote.co.uk/api/beta/postcode/GL12EQ.json is a good example of this: a couple of months ago that might have been returning a `polling_station_known` response, but today it isn't.

Its also problematic if you're trying to create a test case for an esoteric edge case which relies on input data that you don't have examples of in production data.

As soon as you can't just arbitrarily delete the fixture and re-record it, you're into trying to edit a JSON response in a yaml file, I think tools like that start working against you.
